src/XliffTasks/Tasks/UpdateXlf.cs

```diff
@@ -28,38 +30,67 @@ protected override void ExecuteCore()
-                string sourceDocumentId = GetSourceDocumentId(sourcePath);
+                OnlyOneProcessToWorkOn(Path.GetFullPath(item.ItemSpec),
+                    () => {
+                        string sourcePath = item.ItemSpec;
+                        string sourceFormat = item.GetMetadataOrThrow(MetadataKey.XlfSourceFormat);
+                        TranslatableDocument sourceDocument = LoadSourceDocument(sourcePath, sourceFormat);
+                        string sourceDocumentId = GetSourceDocumentId(sourcePath);
 
-                foreach (var language in Languages)
-                {
-                    string xlfPath = GetXlfPath(sourcePath, language);
-                    XlfDocument xlfDocument;
+                        foreach (var language in Languages)
+                        {
+                            string xlfPath = GetXlfPath(sourcePath, language);
+                            XlfDocument xlfDocument;
 
-                    try
-                    {
-                        xlfDocument = LoadXlfDocument(xlfPath, language, createIfNonExistent: AllowModification);
-                    }
-                    catch (FileNotFoundException ex) when (ex.FileName == xlfPath)
-                    {
-                        Release.Assert(!AllowModification);
-                        throw new BuildErrorException($"'{xlfPath}' for '{sourcePath}' does not exist. {HowToUpdate}");
-                    }
+                            try
+                            {
+                                xlfDocument = LoadXlfDocument(xlfPath, language, createIfNonExistent: AllowModification);
+                            }
+                            catch (FileNotFoundException ex) when (ex.FileName == xlfPath)
+                            {
+                                Release.Assert(!AllowModification);
+                                throw new BuildErrorException($"'{xlfPath}' for '{sourcePath}' does not exist. {HowToUpdate}");
+                            }
+
+                            if (!xlfDocument.Update(sourceDocument, sourceDocumentId))
+                            {
+                                continue; // no changes
+                            }
+
+                            if (!AllowModification)
+                            {
+                                throw new BuildErrorException($"'{xlfPath}' is out-of-date with '{sourcePath}'. {HowToUpdate}");
+                            }
+
+                            Directory.CreateDirectory(Path.GetDirectoryName(xlfPath));
+                            xlfDocument.Save(xlfPath);
+                        }
+                    });
+            }
+        }
 
-                    if (!xlfDocument.Update(sourceDocument, sourceDocumentId))
+        private static void OnlyOneProcessToWorkOn(string filePath, Action action)
+        {
+            using (Mutex mtx = new Mutex(false, @"Global\" + filePath.GetHashCode().ToString()))
```

Comment:
This name is incorrect since string.GetHashCode() is only guaranteed to be the same for a given string with the same AppDomain. On desktop, [this] can be enabled. On .NET Core, it is on by default!

Reply:
I don't think this is enough synchronization since one thread could be reading the Xlf to get translations (different task) while another is updating it.

Reply:
I think instead of locking for the whole duration of Updatexlf, we can instead:

1. Use FileShare.None in Document.Save(path) and FileShare.Read in Document.Load(path)

2. Have OpenXlfDocument retry on IOException/UnauthorizedAccessException.

3. Replace xlfDoucment.Save with SaveXlfDocument helper that saves to temp path and does File.Replace in a retry loop to move the file.

This should maximize parallelism and get rid of the need to create 1:1 mutex:file, getting rid of the naming problem and other things we may be missing here such as handling AbandonedMutexException.

On Windows:
  * Readers will prevent the file from being overwritten due to FileShare.Read. 
  * Readers may also be unable to read while writer is replacing the file.
  * Readers can read in parallel, but when they contend with a writer, the retrying kick in to resolve it.

On Unix:
  * FileShare.Read does nothing, but...
  * File.Replace is implemented with rename system call that will mean that even though writers can overwrite while reading is happening, each reader will see file before or after overwrite, not in between, which is the same reader-writer semantic we want.

Reply:
> string.GetHashCode() 

That is a surprise, i am still thinking about ^37 algorithm. Let me spend more time on that